### PR TITLE
Brendancol/geo transfer functions

### PR DESCRIPTION
**Pull request by brendancol**

### CODE REVIEW

**Comment by ceball** [comment]
On travis and appveyor, after installing datashader's dependencies using conda, a develop/editable install of datashader is done using pip (`pip install --no-deps -e .`). The current error you're seeing on travis and appveyor seems to be from pip complaining about an already installed package, and is apparently a problem affecting many projects recently (see https://github.com/conda-forge/testpath-feedstock/issues/7, https://github.com/conda-forge/jupyter-feedstock/issues/5).

I had this problem myself on another project, and fixed it temporarily by running `conda install -c "testpath<0.4"`.

**Comment by jbednar** [comment]
"clearly an installed package with broken metadata shouldn't break pip like this" -- agreed!  @brendancol, please do add that pin to see if it fixes the build or at least lets you go on to the next issue.

File: datashader/geo.py [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+    -------
+    data: DataArray
+    """
+
+    if not isinstance(agg, DataArray):
+        raise TypeError("agg must be instance of DataArray")
+
+    if not agg.attrs.get('res'):
+        #TODO: maybe monkey-patch a "res" attribute valueing unity is reasonable
+        raise ValueError('input xarray must have numeric `res` attr.')
```
Comment (jbednar) [inline datashader/geo.py]: Shouldn't the docstring explain that res is required, and what it should be?

File: datashader/geo.py [hunk 2]
```diff
@@ -0,0 +1,472 @@
...
+                                     data[y-1, x],   data[y, x],   data[y+1, x],
+                                     data[y-1, x+1], data[y, x+1], data[y+1, x+1]]
+                out[y, x] = (a+b+c+d+e+f+g+h+i) / 9
+            else:
+                out[y, x] = data[y, x]
+    return out
+
+def mean(agg, passes=1, excludes=[np.nan]):
+    """
+    Returns Mean filtered array using a 3x3 window sum number of times
```
Comment (jbednar) [inline datashader/geo.py]: "Returns mean filtered array using a 3x3 window applied in multiple passes"?

File: examples/user_guide/8_Geography.ipynb [hunk 3]
```diff
@@ -4,14 +4,317 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Bump\n",
+    "Bump mapping is a cartographic technique often used to create the appearance of trees or other land features.\n",
+    "\n",
+    "The `datashader.bump` will produce a bump aggregate that can then used to add detail to the terrain.  In this case, I will pretend the bumps are trees and shade them with green."
```
Comment (jbednar) [inline examples/user_guide/8_Geography.ipynb]: can then be used

File: datashader/geo.py (hunk 2 above)
Comment (brendancol) [reply datashader/geo.py]: Yes...revised...

File: datashader/geo.py (hunk 1 above)
Comment (brendancol) [reply datashader/geo.py]: I've added the following note to the docstring:

```
    Notes:
    ------

    Input aggregate must contain a `res` (numeric) attribute which specifies the
    x/y cellsize for use in calculating slope.  It is assumed that cells are square
    and have the same resolution in x/y dimensions.

    Algorithm References:
     - http://desktop.arcgis.com/en/arcmap/10.3/tools/spatial-analyst-toolbox/how-slope-works.htm
     - Burrough, P. A., and McDonell, R. A., 1998. Principles of Geographical Information Systems (Oxford University Press, New York), pp 406
```

**Comment by brendancol** [comment]
@jbednar I think I've hit all your review notes.  CI is also now passing.

**Comment by jbednar** [comment]
Thanks!  Now on to the tile generation!

**Pull request merged**